libraries/lib-xml/XMLAttributeValueView.cpp

```diff
@@ -0,0 +1,415 @@
+}
+
+int16_t XMLAttributeValueView::GetInt16(
+   short defaultValue /*= {}*/) const noexcept
+{
+   Get(defaultValue);
+   return defaultValue;
+}
+
+uint16_t XMLAttributeValueView::GetUInt16(
+   unsigned short defaultValue /*= {}*/) const noexcept
+{
+   Get(defaultValue);
+   return defaultValue;
+}
+
+int32_t
+XMLAttributeValueView::GetInt32(int32_t defaultValue /*= {}*/) const noexcept
+{
+   Get(defaultValue);
+   return defaultValue;
+}
+
+uint32_t
+XMLAttributeValueView::GetUInt32(uint32_t defaultValue /*= {}*/) const noexcept
+{
+   Get(defaultValue);
+   return defaultValue;
+}
+
+int64_t
+XMLAttributeValueView::GetInt64(int64_t defaultValue /*= {}*/) const noexcept
+{
+   Get(defaultValue);
+   return defaultValue;
+}
+
+uint64_t
+XMLAttributeValueView::GetUInt64(uint64_t defaultValue /*= {}*/) const noexcept
+{
+   Get(defaultValue);
+   return defaultValue;
+}
+
+float XMLAttributeValueView::GetFloat(
+   float defaultValue /*= {}*/) const noexcept
+{
+   Get(defaultValue);
+   return defaultValue;
+}
+
+double
+XMLAttributeValueView::GetDouble(double defaultValue /*= {}*/) const noexcept
+{
+   Get(defaultValue);
+   return defaultValue;
+}
+
+std::string_view XMLAttributeValueView::GetStringView(
+   std::string_view defaultValue /*= {}*/) const noexcept
+{
+   Get(defaultValue);
+   return defaultValue;
+}
+
+std::string XMLAttributeValueView::ToString() const
+{
+   switch (mType)
+   {
+   case XMLAttributeValueView::Type::Null:
+      return {};
+   case XMLAttributeValueView::Type::SignedInteger:
+      return std::to_string(mInteger);
+   case XMLAttributeValueView::Type::UnsignedInteger:
+      return std::to_string(static_cast<uint64_t>(mInteger));
+   case XMLAttributeValueView::Type::Float:
+      return std::to_string(mFloat);
+   case XMLAttributeValueView::Type::Double:
+      return std::to_string(mDouble);
+   case XMLAttributeValueView::Type::StringView:
+      return std::string(mStringView.Data, mStringView.Length);
+   }
+
+   return {};
+}
+
+wxString XMLAttributeValueView::ToWString() const
+{
+   switch (mType)
+   {
+   case XMLAttributeValueView::Type::Null:
+      return {};
+   case XMLAttributeValueView::Type::SignedInteger:
+      return wxString() << mInteger;
+   case XMLAttributeValueView::Type::UnsignedInteger:
+      return wxString() << mInteger; static_cast<uint64_t>(mInteger);
+   case XMLAttributeValueView::Type::Float:
+      return wxString() << mFloat;
+   case XMLAttributeValueView::Type::Double:
+      return wxString() << mDouble;
+   case XMLAttributeValueView::Type::StringView:
+      return wxString::FromUTF8(mStringView.Data, mStringView.Length);
+   }
+
+   return {};
+}
+
+template <typename ResultType>
+bool CheckInteger(ResultType& output, int64_t value) noexcept
+{
+   constexpr int64_t minValue = std::numeric_limits<ResultType>::min();
+   constexpr int64_t maxValue = std::numeric_limits<ResultType>::max();
+
+   if (minValue <= value && value <= maxValue)
```

Comment:
Maybe cast to unsigned and express this with one bitwise &

Reply:
The compiler replaces it with something like `(value - min) < (max - min)` if the ResultType is smaller than int64_t and simply compiles out this check if ResultType is int64_t.

I feel like the direct comparison here expresses the intention better both for the reader and for the compiler.

Reply:
Cool.  Well at least one compiler.

Reply:
I've checked it over Clang, MSVC, and GCC. All three generated a single comparison instruction here and compiled out the int64_t case.

Reply:
Triple cool.